### PR TITLE
Refactor AliceSpawner

### PR DESCRIPTION
**Pull request by tcharding**
We would like to be able to re-spawn Alice for swaps that are in progress if cnd restarts.  Currently this is not possible because of the way that `AliceSpawner` is implemented.

Refactor `AliceSpawner` so that Alice can be re-spawned.  This is in line with how `BobSpawner` currently is.  This means
- Store the swap in the metadata store and the state store
- Send the request to peer node (Bob)
- Spawn Alice.

Resolves: #1570

### CODE REVIEW

File: cnd/src/comit_client.rs [hunk 1]
```diff
@@ -37,3 +48,117 @@ pub enum RequestError {
...
+    >(
+        &self,
+        dial_information: DialInformation,
+        request: rfc003::Request<AL, BL, AA, BA>,
+    ) -> Box<dyn Future<Item = rfc003::Response<AL, BL>, Error = RequestError> + Send>
+    where
+        LedgerEventDependencies: CreateLedgerEvents<AL, AA> + CreateLedgerEvents<BL, BA>,
+    {
+        let request = build_swap_request(request)
+            .expect("constructing a frame::OutoingRequest should never fail!");
```
Comment (D4nte) [inline cnd/src/comit_client.rs]:
```suggestion
            .expect("constructing a frame::OutgoingRequest should never fail!");
```

File: cnd/src/http_api/mod.rs [hunk 2]
```diff
@@ -306,33 +309,31 @@ impl<'de> Deserialize<'de> for DialInformation {
 /// to another implementation. This allows us to keep the number of arguments to
 /// HTTP API controllers small and still access all the functionality we need.
 #[derive(Debug)]
-pub struct Dependencies<M, S, A, B, N> {
-    pub metadata_store: Arc<M>,
-    pub state_store: Arc<S>,
+pub struct Dependencies<A, B, S> {
+    pub metadata_store: Arc<InMemoryMetadataStore>,
```
Comment (D4nte) [inline cnd/src/http_api/mod.rs]:
```suggestion
    pub metadata_store: Arc<MetadataStore>,
```
I understand that a MedatadaStore will always be in memory. As all Rust struct are :)

Comment (tcharding) [reply cnd/src/http_api/mod.rs]: Commented in another thread on this.  Perhaps we can chat offline about the motivation - basically I am not a big fan of generics when there is only, and only likely ever going to be, a single struct that implements the trait.  It seems to add needless complexity, of course you have much more Rust experience so I am open to your opinions on this.

Comment (D4nte) [reply cnd/src/http_api/mod.rs]: I don't understand your comment.

What I am saying is: Why do we call this `InMemoryMedataStore` when, as per your previous comments, will ever be only one type of `MetadataStore` due to the fact that we always load db items in this `InMemoryMedataStore`.

I am just proposing a rename here. Not sure you why you are going about generics and all.

I don't like the fact that we prefix the name with `InMemory`. If there is not going to be a `OnFilesystem` version of this then let's not prefix it with `InMemory`. Let's rename to `MetadataStore`.

Comment (tcharding) [reply cnd/src/http_api/mod.rs]: Cool, I agree with you totally.  I have it on my list of clean ups to remove the generics and just have a single MetatadataStore.  And do the same for StateStore.

**Comment by tcharding** [comment]
If its ok with the team I will merge https://github.com/comit-network/comit-rs/pull/1604 after review as it is a superset of this one (and has been rebased on top of current master already).  Then I will close this one, leaving open for now however until the other merges.

**Comment by thomaseizinger** [comment]
https://github.com/comit-network/comit-rs/pull/1604 seems to be a strict superset and we decided to merge that. Will close this then!

**Pull request closed without merging**